### PR TITLE
fix types for array access

### LINKED ISSUE #2040: Typescript Error: accessing array entry does not comply to key type
## 🐛 Bug Report

According to what it says here: https://www.i18next.com/translation-function/objects-and-arrays, the following code should be valid in a react+typescript environment,

```tsx
const result1 = i18next.t('arrayJoin', { joinArrays: '+' });
// -> "line1+line2+line3"

const result2 = i18next.t('arrayJoinWithInterpolation', { myVar: 'interpolate', joinArrays: ' ' });
// -> "you can interpolate"

const result3 = i18next.t('arrayOfObjects.0.name');
// -> "tom"
```

but in reality, this:

```tsx
const result3 = i18next.t('arrayOfObjects.0.name');
```

returns a TS error:

```sh
src/App.tsx:12:29 - error TS2345: Argument of type '["arrayOfObjects.0.name"]' is not assignable to parameter of type '[key: "arrayJoin" | TemplateStringsArray | "arrayJoinWithInterpolation" | "arrayOfObjects" | "translation:arrayJoin" | "translation:arrayJoinWithInterpolation" | "translation:arrayOfObjects" | ("arrayJoin" | ... 5 more ... | "translation:arrayOfObjects")[], options?: (TOptionsBase & ... 1 more ... & InterpolationMap...'.
  Type '["arrayOfObjects.0.name"]' is not assignable to type '[key: "arrayJoin" | TemplateStringsArray | "arrayJoinWithInterpolation" | "arrayOfObjects" | "translation:arrayJoin" | "translation:arrayJoinWithInterpolation" | "translation:arrayOfObjects" | ("arrayJoin" | ... 5 more ... | "translation:arrayOfObjects")[], options?: (TOptionsBase & ... 1 more ... & InterpolationMap...'.
    Type at position 0 in source is not compatible with type at position 0 in target.
      Type '"arrayOfObjects.0.name"' is not assignable to type '"arrayJoin" | TemplateStringsArray | "arrayJoinWithInterpolation" | "arrayOfObjects" | "translation:arrayJoin" | "translation:arrayJoinWithInterpolation" | "translation:arrayOfObjects" | ("arrayJoin" | ... 5 more ... | "translation:arrayOfObjects")[]'. Did you mean '"arrayOfObjects"'?

12   const result3 = i18next.t('arrayOfObjects.0.name');
                               ~~~~~~~~~~~~~~~~~~~~~~~
```

## To Reproduce

https://github.com/sreucherand/react-i18next/tree/master/example/react-typescript/simple

```sh
yarn
yarn typechecks
```

### PR DESCRIPTION
**Pull request by Gotos**
The typing for accessing arrays was broken (yielding type errors on any access) and should be fixed with this PR. This should fix #2040 - I hope I spotted all the edge cases.

#### Checklist

- [x] only relevant code is changed (make a diff before you submit the PR)
- [x] run tests `npm run test`
- [x] tests are included
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/i18next/.github/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Comment by adrai** [comment]
@marcalexiei can you review this?
@Gotos can you also adapt the t.v4.d.ts file?

**Comment by Gotos** [comment]
Just to verify: The t4 file is used when using typescript 4 instead of 5? Or what is it for?

**Comment by Gotos** [comment]
I updated the v4 file. I'm not entirely sure how you usually test the v4 files (I didn't see any scripts for that in the package.json) so it's not well tested (yet). On first glance (downgrading my local TS version), it seems to work, but I wasn't able to run the full tests like this.

File: typescript/t.d.ts [hunk 1]
```diff
@@ -51,6 +51,10 @@ type AppendNamespace<Ns, Keys> = `${Ns & string}${_NsSeparator}${Keys & string}`
  ***************************************************** */
 type KeysBuilderWithReturnObjects<Res, Key = keyof Res> = Key extends keyof Res
   ? Res[Key] extends $Dictionary
+    ?
+        | JoinKeys<Key, WithOrWithoutPlural<keyof $OmitArrayKeys<Res[Key]>>>
+        | JoinKeys<Key, KeysBuilderWithReturnObjects<Res[Key]>>
+    : Res[Key] extends any[]
```
Comment (marcalexiei) [inline typescript/t.d.ts:57]:
```suggestion
    : Res[Key] extends unknown[]
```

File: typescript/t.d.ts [hunk 2]
```diff
@@ -59,6 +63,8 @@ type KeysBuilderWithReturnObjects<Res, Key = keyof Res> = Key extends keyof Res
 
 type KeysBuilderWithoutReturnObjects<Res, Key = keyof $OmitArrayKeys<Res>> = Key extends keyof Res
   ? Res[Key] extends $Dictionary
+    ? JoinKeys<Key, KeysBuilderWithoutReturnObjects<Res[Key]>>
+    : Res[Key] extends any[]
```
Comment (marcalexiei) [inline typescript/t.d.ts:67]:
```suggestion
    : Res[Key] extends unknown[]
```

File: typescript/t.v4.d.ts [hunk 3]
```diff
@@ -51,6 +51,10 @@ type AppendNamespace<Ns, Keys> = `${Ns & string}${_NsSeparator}${Keys & string}`
  ***************************************************** */
 type KeysBuilderWithReturnObjects<Res, Key = keyof Res> = Key extends keyof Res
   ? Res[Key] extends $Dictionary
+    ?
+        | JoinKeys<Key, WithOrWithoutPlural<keyof $OmitArrayKeys<Res[Key]>>>
+        | JoinKeys<Key, KeysBuilderWithReturnObjects<Res[Key]>>
+    : Res[Key] extends any[]
```
Comment (marcalexiei) [inline typescript/t.v4.d.ts:57]:
```suggestion
    : Res[Key] extends unknown[]
```

File: typescript/t.v4.d.ts [hunk 4]
```diff
@@ -169,7 +175,11 @@ type ParseTReturn<
     ? ParseTReturnPluralOrdinal<Res, Key>
     : ParseTReturnPlural<Res, Key>
   : // otherwise access plain key without adding plural and ordinal suffixes
-    Res[Key & keyof Res];
+  Res extends any[]
```
Comment (marcalexiei) [inline typescript/t.v4.d.ts:178]:
```suggestion
  Res extends unknown[]
```

**Review by marcalexiei — CHANGES_REQUESTED** [review]
Just a small change: let's avoid `any[]`, `unknown[]` should be better and do the same.
I marked all the occurrences I was able to find.

Everything else looks good to me!

**Comment by adrai** [comment]
included in v23.7.14
